util/deephash/deephash.go

```diff
@@ -30,14 +30,19 @@ type hasher struct {
 	h       hash.Hash
 	bw      *bufio.Writer
 	scratch [scratchSize]byte
-	visited map[uintptr]bool
+
+	// visitStack is a stack of pointers visited.
+	// Pointers are pushed onto the stack when visited, and popped when leaving.
+	// The integer value is the depth at which the pointer was visited.
+	// The length of this stack should be zero after ever hashing operation.
+	visitStack map[uintptr]int
```

Comment:
Using a uintptr as the key here will break if there's a moving GC. Maybe add:

```go
import _ "go4.org/unsafe/assume-no-moving-gc"
```

Pointers change today on stack copies, but I think we're saved from that because anything that gets passed to a hasher will get heap-allocated today. But I'm a bit nervous that future escape analysis improvements could change that, in which case recursive print calls could cause the stack to grow, thereby breaking cycle detection.

I'd feel more comfortable if we used unsafe.Pointer here so we didn't have to worry about that, assuming we can find a safe way to get an unsafe.Pointer out of package reflect.

Reply:
I prefer this be an `unsafe.Pointer`, but I can understand the aversion towards using `unsafe`.

Reply:
I'd like this to be an unsafe.Pointer too. Sorry if that didn't come through clearly above. @bradfitz you OK with that?

Reply:
yeah, I'm fine with that.

Reply:
Alright, I switched it to using `unsafe.Pointer`. I created a `pointer` wrapper type to reduce that number of references to `unsafe.Pointer` and document that we only depend on it for comparability and not something more sketchy.